src/async_impl/multipart.rs

```diff
@@ -0,0 +1,390 @@
+//! multipart/form-data
+use std::borrow::Cow;
+use std::fmt;
+
+use mime_guess::Mime;
+use url::percent_encoding::{self, EncodeSet, PATH_SEGMENT_ENCODE_SET};
+use uuid::Uuid;
+use http::HeaderMap;
+
+use futures::Stream;
+
+use super::Body;
+
+/// An async multipart/form-data request.
+pub struct Form {
+    boundary: String,
+    fields: Vec<(Cow<'static, str>, Part)>,
+    percent_encoding: PercentEncoding,
+}
+
+enum PercentEncoding {
+    PathSegment,
+    AttrChar,
+}
+
+impl Form {
+    /// Creates a new async Form without any content.
+    pub fn new() -> Form {
+        Form {
+            boundary: format!("{}", Uuid::new_v4().to_simple()),
+            fields: Vec::new(),
+            percent_encoding: PercentEncoding::PathSegment,
+        }
+    }
+
+    /// Get the boundary that this form will use.
+    #[inline]
+    pub fn boundary(&self) -> &str {
+        &self.boundary
+    }
+
+    /// Add a data field with supplied name and value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let form = reqwest::async::multipart::Form::new()
+    ///     .text("username", "seanmonstar")
+    ///     .text("password", "secret");
+    /// ```
+    pub fn text<T, U>(self, name: T, value: U) -> Form
+    where T: Into<Cow<'static, str>>,
+          U: Into<Cow<'static, str>>,
+    {
+        self.part(name, Part::text(value))
+    }
+
+    /// Adds a customized Part.
+    pub fn part<T>(mut self, name: T, part: Part) -> Form
+    where T: Into<Cow<'static, str>>,
+    {
+        self.fields.push((name.into(), part));
+        self
+    }
+
+    /// Configure this `Form` to percent-encode using the `path-segment` rules.
+    pub fn percent_encode_path_segment(mut self) -> Form {
+        self.percent_encoding = PercentEncoding::PathSegment;
+        self
+    }
+
+    /// Configure this `Form` to percent-encode using the `attr-char` rules.
+    pub fn percent_encode_attr_chars(mut self) -> Form {
+        self.percent_encoding = PercentEncoding::AttrChar;
+        self
+    }
+
+    /// Consume this instance and transform into an instance of hyper::Body for use in a request.
+    pub(crate) fn stream(mut self) -> hyper::Body {
+      if self.fields.len() == 0 {
```

Comment:
To match the rest of the repo, mind making the indents use 4 spaces? Sorry for the nitpick,,,

Reply:
Lol no you're good, I'll make that change.